url_shortener/views/handlers.py

```diff
@@ -1,60 +1,332 @@
+        })
+
+    if id is None or long_link is None:
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Incorrect key.'
+        })
 
+    return HTTPFound(location=long_link)
+
+
+def create_shortlink(request: Request) -> Response:
+    """ Creates a shortlink id, by using id if provided or generating 4 char random string."""
+    logic: Logic = request.registry.logic
+
+    try:
+        url_provided = request.json_body.get('url')
+        id_provided = request.json_body.get('id')
+    except Exception as e:
+        print(e)
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Request body is missing.'
+        })
+
+    if url_provided is None:
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'URL must be provided.'
+        })
+
+    # If custom id is not provided, or this id already exists, we give an random one.
+    while id_provided is None or logic.read_by_key(id_provided) is not None:
+        id_provided = ''.join(random.choice(string.ascii_uppercase + string.ascii_lowercase) for _ in range(4))
+
+    # People shouldn't overwrite user id's so we'll block it.
+    if any(c.isdigit() for c in id_provided):
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Numbers in id is not allowed.'
+        })
+
+    # URL check if it's valid. (stolen from django url validation regex)
+    regex_check = re.compile(
+        r'^(?:http|ftp)s?://'  # http:// or https://
+        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # domain...
+        r'localhost|'  # localhost...
+        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'  # ...or ip
+        r'(?::\d+)?'  # optional port
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+
+    if not re.match(regex_check, url_provided.lower()) is not None:
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'URL is incorrect.'
+        })
+
+    success = logic.save_if_not_exists(id_provided, url_provided.lower())
+
+    # If saving the shortlink succeeds, save this info to user.
     if success:
-        return Response(
-            status=httplib.CREATED,
-            headerlist=[('Location',
-                         request.registry.base_url + '/resource/' + key)],
-        )
+        new_user = logic.find_user_by_token(get_auth_token(request))
+        new_user.urls.append(id_provided)
+        logic.save_user(str(new_user.id), new_user)
+        return Response(status=httplib.OK, json_body={
+            'status': 'success',
+            'description': 'Successfully created short url.',
+            'id': id_provided,
+            'shortened_link': domain_name + id_provided
```

Comment:
Instead of defining `domain_name` constant here, you can read `request.registry.base_url`